New version 2.5.1 introduces multiple OrphanContainerBuildError

## Description

The `master` branch build of the `AllenSDK` project has has experienced a number of test fauilures introduced by the latest version of `hdmf`. All errors seem to be `hdmf.build.errors.OrphanContainerBuildError`s

One particular example error message:
```
hdmf.build.errors.OrphanContainerBuildError: EyeTracking (EyeTracking): Linked EllipseSeries 'eye_tracking' has no parent. Remove the link or ensure the linked container is added properly.
```

<!--Please provide the following in your bug report:-->

## Steps to Reproduce

Here is a condensed version of the Bamboo build failure logs
```
    def test_nwb_end_to_end(tmpdir_factory):    
         d1 = BehaviorOphysExperiment.from_lims(oeid)
>       BehaviorOphysNwbApi(nwb_filepath).save(d1)

```
```
self = <pynwb.io.core.NWBContainerMapper object at 0x7fd444a0b470>
builder = EyeTracking GroupBuilder {'attributes': {}, 'groups': {}, 'datasets': {}, 'links': {}}
spec = {'doc': 'eye tracking', 'name': 'eye_tracking', 'neurodata_type_inc': 'EllipseSeries'}
value = eye_tracking abc.EllipseSeries at 0x140551993871664
Fields:
  comments: no comments
  conversion: 1.0
  data: [[323.36...320000e-01 2.53310000e-01 ... 4.52830716e+03
 4.52834063e+03 4.52837453e+03]
  timestamps_unit: seconds
  unit: meters

build_manager = <hdmf.build.manager.BuildManager object at 0x7fd5054939b0>
source = '/tmp/pytest-of-root/pytest-0/popen-gw3/test_nwb_end_to_end0/nwbfile.nwb'
parent_container = EyeTracking abc.EllipseEyeTracking at 0x140552893882096
Fields:

export = False

    def __add_containers(self, builder, spec, value, build_manager, source, parent_container, export):
        if isinstance(value, AbstractContainer):
            self.logger.debug("    Adding container %s '%s' with parent %s '%s' to %s '%s'"
                              % (value.__class__.__name__, value.name,
                                 parent_container.__class__.__name__, parent_container.name,
                                 builder.__class__.__name__, builder.name))
            if value.parent is None:
                if (value.container_source == parent_container.container_source or
                        build_manager.get_builder(value) is None):
                    # value was removed (or parent not set) and there is a link to it in same file
                    # or value was read from an external link
>                   raise OrphanContainerBuildError(builder, value)
E                   hdmf.build.errors.OrphanContainerBuildError: EyeTracking (EyeTracking): Linked EllipseSeries 'eye_tracking' has no parent. Remove the link or ensure the linked container is added properly.

/opt/conda/envs/py36/lib/python3.6/site-packages/hdmf/build/objectmapper.py:1049: OrphanContainerBuildError

```

## Environment

<!--Please describe your environment according to the following bullet points.-->

    Python Executable: Conda or Python
    Python Version: Python 3.6, 3.7
    Operating System: Linux
    HDMF Version: 2.5.1

## Checklist

- [ ] Have you ensured the bug was not already [reported](https://github.com/hdmf-dev/hdmf/issues) ?
- [ x ] Have you included a brief and descriptive title?
- [ x ] Have you included a clear description of the problem you are trying to solve?
- [ x ] Have you included a minimal code snippet that reproduces the issue you are encountering?
- [ ] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst) document?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,9 +1,11 @@
 # HDMF Changelog
 
-## Upcoming
+## HDMF 2.5.2 (May 11, 2021)
 
 ### Bug fix
 - Add explicit `setuptools` requirement. @hrnciar (#596)
+- Fix issue with generated custom classes that use a custom fields name (e.g., PyNWB uses `__nwbfields__` instead
+  of `__fields__`). @rly (#598)
 
 ## HDMF 2.5.1 (April 23, 2021)
 

--- a/src/hdmf/build/classgenerator.py
+++ b/src/hdmf/build/classgenerator.py
@@ -208,7 +208,7 @@ class CustomClassGenerator:
     @classmethod
     def process_field_spec(cls, classdict, docval_args, parent_cls, attr_name, not_inherited_fields, type_map, spec):
         """Add __fields__ to the classdict and update the docval args for the field spec with the given attribute name.
-        :param classdict: The dict to update with __fields__.
+        :param classdict: The dict to update with __fields__ (or a different parent_cls._fieldsname).
         :param docval_args: The list of docval arguments.
         :param parent_cls: The parent class.
         :param attr_name: The attribute name of the field spec for the container class to generate.
@@ -224,7 +224,7 @@ class CustomClassGenerator:
             fields_conf['child'] = True
         # if getattr(field_spec, 'value', None) is not None:  # TODO set the fixed value on the class?
         #     fields_conf['settable'] = False
-        classdict.setdefault('__fields__', list()).append(fields_conf)
+        classdict.setdefault(parent_cls._fieldsname, list()).append(fields_conf)
 
         docval_arg = dict(
             name=attr_name,
@@ -262,15 +262,15 @@ class CustomClassGenerator:
     @classmethod
     def post_process(cls, classdict, bases, docval_args, spec):
         """Convert classdict['__fields__'] to tuple and update docval args for a fixed name and default name.
-        :param classdict: The class dictionary to convert.
+        :param classdict: The class dictionary to convert with '__fields__' key (or a different bases[0]._fieldsname)
         :param bases: The list of base classes.
         :param docval_args: The dict of docval arguments.
         :param spec: The spec for the container class to generate.
         """
         # convert classdict['__fields__'] from list to tuple if present
-        fields = classdict.get('__fields__')
+        fields = classdict.get(bases[0]._fieldsname)
         if fields is not None:
-            classdict['__fields__'] = tuple(fields)
+            classdict[bases[0]._fieldsname] = tuple(fields)
 
         # if spec provides a fixed name for this type, remove the 'name' arg from docval_args so that values cannot
         # be passed for a name positional or keyword arg

--- a/tests/unit/build_tests/test_classgenerator.py
+++ b/tests/unit/build_tests/test_classgenerator.py
@@ -783,7 +783,7 @@ class TestBaseProcessFieldSpec(TestCase):
         )
 
         classdict = {}
-        bases = []
+        bases = [Container]
         docval_args = [{'name': 'name', 'type': str, 'doc': 'name'},
                        {'name': 'attr1', 'type': ('array_data', 'data'), 'doc': 'a string attribute',
                         'shape': [None]}]
@@ -810,7 +810,7 @@ class TestBaseProcessFieldSpec(TestCase):
         )
 
         classdict = {}
-        bases = []
+        bases = [Container]
         docval_args = [{'name': 'name', 'type': str, 'doc': 'name'},
                        {'name': 'attr1', 'type': ('array_data', 'data'), 'doc': 'a string attribute',
                         'shape': [None]}]

